AdyenCard/Components/Card/CardPublicKeyConsumer.swift

```diff
@@ -13,32 +13,19 @@ internal protocol CardPublicKeyConsumer: PaymentComponent {
 
 extension CardPublicKeyConsumer {
     internal typealias CardKeySuccessHandler = (_ cardPublicKey: String) -> Void
-    internal typealias CardKeyFailureHandler = (_ error: Swift.Error) -> Void
 
-    internal func fetchCardPublicKey(onError: CardKeyFailureHandler? = nil, completion: @escaping CardKeySuccessHandler) {
-        do {
-            try cardPublicKeyProvider.fetch { [weak self] in
-                self?.handle(result: $0, onError: onError, completion: completion)
-            }
-        } catch {
-            if let onError = onError {
-                onError(error)
-            } else {
-                delegate?.didFail(with: error, from: self)
-            }
+    internal func fetchCardPublicKey(completion: @escaping CardKeySuccessHandler) {
+        cardPublicKeyProvider.fetch { [weak self] in
+            self?.handle(result: $0, completion: completion)
         }
     }
 
-    private func handle(result: Result<String, Swift.Error>, onError: CardKeyFailureHandler? = nil, completion: CardKeySuccessHandler) {
+    private func handle(result: Result<String, Swift.Error>, completion: CardKeySuccessHandler) {
         switch result {
         case let .success(key):
             completion(key)
         case let .failure(error):
-            if let onError = onError {
-                onError(error)
-            } else {
-                delegate?.didFail(with: error, from: self)
-            }
+            delegate?.didFail(with: error, from: self)
```

Comment:
maybe as a parameter to `fetchCardPublicKey` we could pass a handler of type `Result<String, Swift.Error>` and use it for both success and fail scenarios, instead of calling `delegate?.didFail(with: error, from: self)`. this way we can remove inheritance of `PaymentComponent` from `protocol CardPublicKeyConsumer`. i think it is a bit better for interface segregation. also, when `submitEncryptedCardData` is called, both success and fail handlers would be in the same place.
```swift
internal protocol CardPublicKeyConsumer: AnyObject {
    var cardPublicKeyProvider: AnyCardPublicKeyProvider { get set }
}

extension CardPublicKeyConsumer {
    internal typealias CardKeyFetchResultHandler = (_ result: Result<String, Swift.Error>) -> Void

    internal func fetchCardPublicKey(completion: @escaping CardKeyFetchResultHandler) {
        cardPublicKeyProvider.fetch(completion: completion)
    }
}

extension CardComponent {
    
    internal func didSelectSubmitButton() {
        guard cardViewController.validate() else {
            return
        }
        
        cardViewController.startLoading()

        fetchCardPublicKey { [weak self] result in
            guard let weakSelf = self else { return }
            
            switch result {
            case .success(let key):
                weakSelf.submitEncryptedCardData(cardPublicKey: key)
            case .failure(let error):
                weakSelf.delegate?.didFail(with: error, from: weakSelf)
            }
        }
    }
```

Reply:
what's the point then of `CardPublicKeyConsumer`? :-D , this is a convenient way instead of having the switch statement in every component that conforms to `CardPublicKeyConsumer`, right?

Reply:
that's kind of my point. `fetchCardPublicKey` is supposed to only fetch the key, and report the result. right now it also does error reporting via `delegate`, which it should not even know about. if you look at apple APIs, completion handler usually is called both in case of success and failure. `CardPublicKeyConsumer` does not need to inherit from `PaymentComponent`, because when inheriting it gets access to a lot of things that it shouldn't (`delegate`, `paymentMethod`, `payment` etc. those are not related to fetching card keys).  I honestly don't think we need this extension, because it consists of one method with almost no logic, and is not reused anywhere. Just calling it directly from the `CardComponent` and handling the result in a switch would be better imho.